ctrd/client.go

```diff
@@ -166,3 +185,125 @@ func (c *Client) Version(ctx context.Context) (containerd.Version, error) {
 
 	return cli.client.Version(ctx)
 }
+
+func (c *Client) runContainerdDaemon(homeDir string, copts clientOpts) error {
+	if homeDir == "" {
+		return fmt.Errorf("ctrd: containerd home dir should not be empty")
+	}
+
+	containerdPath, err := exec.LookPath(copts.containerdBinary)
+	if err != nil {
+		return fmt.Errorf("failed to find containerd binary %s: %v", copts.containerdBinary, err)
+	}
+
+	pidFileName := path.Join(homeDir, "containerd/state", containerdPidFileName)
+	f, err := os.OpenFile(pidFileName, os.O_RDWR|os.O_CREATE, 0600)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+
+	buf := make([]byte, 8)
+	num, err := f.Read(buf)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	if num > 0 {
+		pid, err := strconv.ParseUint(string(buf[:num]), 10, 64)
+		if err != nil {
+			return err
+		}
+		if utils.IsProcessAlive(int(pid)) {
+			logrus.Infof("ctrd: previous instance of containerd still alive (%d)", pid)
+			c.daemonPid = int(pid)
+			return nil
+		}
+	}
+
+	// empty container pid file
+	_, err = f.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	if err := f.Truncate(0); err != nil {
+		return err
+	}
+
+	// if socket file exists, delete it.
+	if _, err := os.Stat(c.socketPath); err == nil {
+		os.RemoveAll(c.socketPath)
+	}
+
+	// Start a new containerd instance
+	args := []string{
+		"-a", copts.rpcAddr,
+		"--root", path.Join(homeDir, "containerd/root"),
+		"--state", path.Join(homeDir, "containerd/state"),
+		"-l", utils.If(copts.debugLog, "debug", "info").(string),
+	}
+
+	cmd := exec.Command(containerdPath, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true, Pdeathsig: syscall.SIGKILL}
+	cmd.Env = nil
+	// clear the NOTIFY_SOCKET from the env when starting containerd
+	for _, e := range os.Environ() {
+		if !strings.HasPrefix(e, "NOTIFY_SOCKET") {
+			cmd.Env = append(cmd.Env, e)
+		}
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	logrus.Infof("ctrd: new containerd process, pid: %d", cmd.Process.Pid)
+
+	if err := utils.SetOOMScore(cmd.Process.Pid, -17); err != nil {
```

Comment:
I remember the `containerd` has the OOM score settings in the configuration. In my opinion, it's better if we maintain configuration for the pouchd. The configuration can help us to locate the problem in the future. How do you think?

Reply:
Agree, reasonable !!!